src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs

```diff
@@ -0,0 +1,166 @@
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Diagnostics;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Servers
+{
+    internal interface IRoundTripTimeMonitor : IDisposable
+    {
+        TimeSpan Average { get; }
+        void AddSample(TimeSpan roundTripTime);
+        void Reset();
+        Task RunAsync();
+    }
+
+    internal class RoundTripTimeMonitor : IRoundTripTimeMonitor
+    {
+        private readonly ExponentiallyWeightedMovingAverage _averageRoundTripTimeCalculator = new ExponentiallyWeightedMovingAverage(0.2);
+
+        private readonly CancellationToken _cancellationToken;
+        private readonly IConnectionFactory _connectionFactory;
+        private bool _disposed;
+        private readonly EndPoint _endPoint;
+        private IConnection _roundTripTimeConnection;
+        private readonly ServerId _serverId;
+        private readonly TimeSpan _heartbeatFrequency;
+        private readonly object _lock = new object();
+        private readonly object _disposeLock = new object();
+
+        public RoundTripTimeMonitor(
+            IConnectionFactory connectionFactory,
+            ServerId serverId,
+            EndPoint endpoint,
+            TimeSpan heartbeatFrequency,
+            CancellationToken cancellationToken)
+        {
+            _cancellationToken = cancellationToken;
+            _connectionFactory = Ensure.IsNotNull(connectionFactory, nameof(connectionFactory));
+            _endPoint = Ensure.IsNotNull(endpoint, nameof(endpoint));
+            _heartbeatFrequency = heartbeatFrequency;
+            _serverId = Ensure.IsNotNull(serverId, nameof(serverId));
+        }
+
+        public TimeSpan Average
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _averageRoundTripTimeCalculator.Average;
+                }
+            }
+        }
+
+        // public methods
+        public void Dispose()
+        {
+            _disposed = true;
+
+            IConnection toDispose;
+            lock (_disposeLock)
+            {
+                toDispose = _roundTripTimeConnection;
+                _roundTripTimeConnection = null;
+            }
+
+            try
+            {
+                toDispose?.Dispose();
+            }
+            catch
+            {
+                // ignore
+            }
+        }
+
+        public async Task RunAsync()
+        {
+            await Task.Yield(); // return control immediately
```

Comment:
>is it because we want control to return to ServerMonitor.Initialize() as soon as possible

exactly. It affects some tests

Reply:
I'm resolving this because I'm happy with the extra explanation posed in the initial comment :) If we're concerned about race conditions between the Tasks, we could start a new comment thread :) @rstam